### PR TITLE
node, joins, and relationships

### PR DESCRIPTION
**Pull request by tannernelson**
Adds `Node` for data conversions alongside joins and relationships.

An example of a full Fluent model in 0.8:

### Base Entity with Node

```swift
import Fluent

final class Atom: Entity {
    var id: Node?
    var name: String
    var protons: Int

    init(name: String, protons: Int) {
        id = nil
        self.name = name
        self.protons = protons
    }

    init(with node: Node, in context: Context) throws {
        id = try node.extract("id")
        name = try node.extract("name")
        protons = try node.extract("protons")
    }

    func makeNode() throws -> Node {
        return try Node([
            "id": id,
            "name": name,
            "protons": protons
        ])
    }
}
```

### Relations as Computed Properties

```swift
extension Atom {
    // Belongs To
    func group() throws -> Parent<Group> {
        return try parent(groupId)
    }

    // Has Many
    func protons() throws -> Children<Proton> {
        return children()
    }

    // Many To Many
    func compounds() throws -> Siblings<Compound> {
        return try siblings()
    }

    // Optionally, Has One
    func nucleus() throws -> Nucleus? {
        return try children().first()
    }
}
```

### Relationship usage

```swift
guard let atom = try Atom.query().filter("name", "Hydrogen").first() else { return }

let group = try atom.group().get()
try atom.group().delete()

let protons = try atom.protons().all()

let heavyCompounds = try atom.compounds().filter("weight", .greaterThan, 500).all()
let allCompounds = try atom.compounds().all()
```

### Custom relationship usage

```swift
guard let atom = try Atom.query().filter("name", "Hydrogen").first() else { return }

let compounds = atom.siblings(Compound.self).all()
```

### Unions

```swift
let query = try Atom.query().union(Compound.self).all()

let query = try Atom.query()
    .union(Compound.self, localKey: "#local_key", foreignKey: "#foreign_key")
    .filter("protons", .greaterThan, 5)
    .filter(Compound.self, "atoms", .lessThan, 128)
    .all()
```

```SQL
SELECT * FROM `atoms`
JOIN `compounds` ON `atoms`.`#local_key` = `compounds`.`#foreign_key`
WHERE ...
```

### Preparations

```swift
extension Atom {
    static func prepare(_ database: Fluent.Database) throws {
        try database.create(entity) { builder in
            builder.id()
            builder.string("name")
            builder.int("protons")
        }
    }
    static func revert(_ database: Fluent.Database) throws {
        try database.delete(entity)
    }
}
```

### CODE REVIEW

File: Sources/Fluent/Utilities/Fluent+Node.swift [hunk 1]
```diff
@@ -0,0 +1,93 @@
+@_exported import Node
+
+public enum ExtractionError: Error {
+    case notDictionary
+    case invalidType
+}
+
+extension Node {
+    public func extract(_ key: String) throws -> [UInt8] {
```
Comment (LoganWright) [inline Sources/Fluent/Utilities/Fluent+Node.swift]: I think for now, all of these extracts can be condensed a bit:

```
public func extract<N: NodeInitializable>(_ key: PathIndex) throws -> N {
    guard let node = self[key] else { throw "unexpected nil" }
    return try N(with: node)
}


public func extract<N: NodeInitializable>(_ key: PathIndex) throws -> [N] {
    guard let node = self[key].nodeArray else { throw "unexpected nil" }
    return try [N](with: node)
}
```

File: Sources/Fluent/Database/Driver.swift [hunk 2]
```diff
@@ -21,11 +21,29 @@ public protocol Driver {
...
+        This allows Fluent extensions to be written that
+        can support custom querying behavior.
+    */
+    @discardableResult
+    func raw(_ raw: String, _ values: [Node]) throws -> Node
+}
+
+extension Driver {
+    @discardableResult
+    public func raw(_ raw: String, _ values: [NodeRepresentable] = []) throws -> Node {
```
Comment (LoganWright) [inline Sources/Fluent/Database/Driver.swift]: do we also want non-homogenous generics here?

ie: `raw<N: NodeRepresentable>(...... values: [N] ...`

File: Sources/Fluent/Entity/Entity.swift [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+/**
+    Represents an entity that can be
+    stored and retrieved from the `Database`.
+*/
+public protocol Entity: Preparation, NodeConvertible {
+    /**
+        The collection or table name
+        for this entity.
+    */
+    static var entity: String { get }
```
Comment (LoganWright) [inline Sources/Fluent/Entity/Entity.swift]: it's mildly disorienting to see entity return String and not `Entity`, maybe this should be `name` or `table`? that's appears to be what's specified in the comment.

File: Sources/Fluent/Query/Filter.swift [hunk 4]
```diff
@@ -4,19 +4,113 @@
...
+        filters.
+
+        Used for filtering results based on whether
+        a result's value is or is not in a set.
+    */
+    @discardableResult
+    public func filter(
+        _ field: String,
+        _ scope: Filter.Scope,
+        _ set: [NodeRepresentable]
```
Comment (LoganWright) [inline Sources/Fluent/Query/Filter.swift]: For collections, we should probably do the generic version here as well. I would say it's more common for collections to be homogenous. We should provide overloads until Swift allows those to be inferred as protocols.

File: Sources/Fluent/Query/Query.swift [hunk 5]
```diff
@@ -49,50 +56,82 @@ public class Query<T: Model> {
...
+        if case .array(let array) = try raw() {
+            for result in array {
+                do {
+                    var model = try T(with: result)
+                    if case .object(let dict) = result {
+                        model.id = dict[database.driver.idKey]
+                    }
+                    models.append(model)
+                } catch {
+                    // FIXME: do what when init fails?
```
Comment (LoganWright) [inline Sources/Fluent/Query/Query.swift]: is it possible for us to propogate up instead of catching or is that not the right behavior?

Comment (tannernelson) [inline Sources/Fluent/Query/Query.swift]: I think we need something more sophisticated than just throwing it up the chain. Imagine getting 1,000 rows from a DB and one doesn't get quite the data it needs so your whole call is disrupted. 

Maybe we allow a closure to be passed for catching and ignoring errors, but besides that we propagate them.

File: Sources/Fluent/Query/Filter.swift (hunk 4 above)
Comment (tannernelson) [inline Sources/Fluent/Query/Filter.swift]: I don't think we're guaranteed they're going to be homogenous. 

```swift
Atom.query().filter("id", .in, [1, "2", [0x33].string])
```

File: Sources/Fluent/Entity/Entity.swift (hunk 3 above)
Comment (tannernelson) [inline Sources/Fluent/Entity/Entity.swift]: Would `static var id: String` be better?

File: Sources/Fluent/Database/Driver.swift (hunk 2 above)
Comment (tannernelson) [inline Sources/Fluent/Database/Driver.swift]: Yes, definitely.

```
let driver = MySQLDriver()
try driver.raw("UPDATE users SET age = ? WHERE name = ?", [5, "foo"])

**Pull request merged**